frontend/apps/reader/modules/readertoc.lua

```diff
@@ -275,147 +280,174 @@ function ReaderToc:getTocTitleOfCurrentPage()
-            end
-        else
-            local depth
-            if level > 0 then
-                depth = level
-            else
-                depth = self:getMaxDepth() + level + 1
-            end
-            for _, v in ipairs(self.toc) do
-                if v.depth == depth then
-                    table.insert(ticks, v.page)
-                end
+        -- Start by building a simple hierarchical ToC tick table
+        for _, v in ipairs(self.toc) do
+            if not self.ticks[v.depth] then
+                self.ticks[v.depth] = {}
             end
+            table.insert(self.ticks[v.depth], v.page)
+        end
+
+        -- Normally the ticks are already sorted, but in rare cases,
+        -- ToC nodes may be not in ascending order
+        for k, _ in ipairs(self.ticks) do
+            table.sort(self.ticks[k])
         end
-        -- normally the ticks are sorted already but in rare cases
-        -- toc nodes may be not in ascending order
-        table.sort(ticks)
-        -- cache ticks only if ticks are available
-        self.ticks[level] = ticks
     end
-    return ticks
+
+    if level then
+        return self.ticks[level]
+    else
+        return self.ticks
+    end
 end
 
-function ReaderToc:getTocTicksForFooter()
+--[[
+Returns a flattened list of ToC ticks, without duplicates
+]]
+function ReaderToc:getTocTicksFlattened()
+    if self.ticks_flattened then return self.ticks_flattened end
+
+    -- It hasn't been cached yet, compute it.
+    local ticks = self:getTocTicks()
     local ticks_candidates = {}
-    local max_level = self:getMaxDepth()
-    for i = 0, -max_level, -1 do
-        local ticks = self:getTocTicks(i)
-        table.insert(ticks_candidates, ticks)
+
+    for _, v in ipairs(ticks) do
+        for depth, page in ipairs(v) do
+            table.insert(ticks_candidates, page)
+        end
     end
-    if #ticks_candidates > 0 then
-        -- Find the finest toc ticks by sorting out the largest one
-        table.sort(ticks_candidates, function(a, b) return #a > #b end)
-        return ticks_candidates[1]
+
+    -- Then drop duplicates (by using the value as a key to make a quick check) (c.f., https://stackoverflow.com/a/20067270)
+    local hash = {}
+    local ticks_flattened = {}
+    for _, v in ipairs(ticks_candidates) do
+        if (not hash[v]) then
+            ticks_flattened[#ticks_flattened+1] = v
+            hash[v] = true
+        end
     end
-    return {}
+
+    -- And finally, sort it again
+    table.sort(ticks_flattened)
+
+    self.ticks_flattened = ticks_flattened
+    return self.ticks_flattened
 end
 
-function ReaderToc:getNextChapter(cur_pageno, level)
-    local ticks = self:getTocTicks(level)
+function ReaderToc:getNextChapter(cur_pageno)
+    local ticks = self:getTocTicksFlattened()
     local next_chapter = nil
-    for i = 1, #ticks do
-        if ticks[i] > cur_pageno then
-            next_chapter = ticks[i]
+    for _, v in ipairs(ticks) do
```

Comment:
(As value is a page, I would have named it p: `for _, p in ipairs(ticks) do if p > cur_pageno then` is a tad more readable/hinting.)